proofs/src/plonk/trash/prover.rs

```diff
@@ -66,13 +70,53 @@ impl<F: WithSmallOrderMulGroup<3> + Ord> Argument<F> {
     }
 }
 
+impl<F: WithSmallOrderMulGroup<3>> PartiallyEvaluated<F> {
+    pub(crate) fn expressions<'a>(
+        &'a self,
+        argument: &'a Argument<F>,
+        trash_challenge: F,
+        advice_evals: &[F],
+        fixed_evals: &[F],
+        instance_evals: &[F],
+        challenges: &[F],
+    ) -> impl Iterator<Item = F> + 'a {
+        let evaluate_expression = |expr: &Expression<F>| {
+            expr.evaluate(
+                &|scalar| scalar,
+                &|_| panic!("virtual selectors are removed during optimization"),
+                &|query| fixed_evals[query.index.unwrap()],
+                &|query| advice_evals[query.index.unwrap()],
+                &|query| instance_evals[query.index.unwrap()],
+                &|challenge| challenges[challenge.index()],
+                &|a| -a,
+                &|a, b| a + &b,
+                &|a, b| a * &b,
+                &|a, scalar| a * &scalar,
+            )
+        };
+
+        let compressed_expressions = (argument.constraint_expressions.iter())
+            .map(evaluate_expression)
+            .fold(F::ZERO, |acc, eval| acc * &trash_challenge + &eval);
+
+        let q = evaluate_expression(argument.selector());
+        vec![compressed_expressions - (F::ONE - q) * self.trash_eval].into_iter()
+    }
+}
+
```

Comment:
This code is duplicate of the verifier. Try to avoid code duplication. Given that both the prover and verifier will use this logic, maybe it makes sense to have it in trash.rs (same for lookup and permutation)